GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -67,8 +96,27 @@ public bool TryInvoke<TResult>(Func<LibGit2Repo, TResult> function, out TResult
 
         private LibGit2Repo GetRepoFromPool()
         {
+            this.ResetRepoDisposalTimer();
+
             LibGit2Repo repo;
-            if (this.pool.TryTake(out repo, TryTakeTimeoutMilliseconds))
+            if (this.pool.TryTake(out repo, TryTakeNoWait))
+            {
+                return repo;
+            }
+
+            if (this.numAvailableRepoAllocations > 0)
+            {
+                if (Interlocked.Decrement(ref this.numAvailableRepoAllocations) >= 0)
+                {
+                    return this.createRepo();
+                }
+                else
+                {
+                    Interlocked.Increment(ref this.numAvailableRepoAllocations);
+                }
+            }
+
+            if (this.pool.TryTake(out repo, TryTakeWaitForever))

```

Comment:
It looks like we a possible race conditions here if `LibGit2RepoPool` were ever created with a size of 1.

As best I can tell we'll always have at least 2 today, but I wanted to mention it and get your thoughts as to whether we should fix it (or update the code to require a minimum size of 2).

Scenario:

1) Thread 1 calls `GetRepoFromPool` and `numAvailableRepoAllocations` is 0
2) Timer thread files, and `TryDropARepo` has not yet called `TryTake`
3) Thread 2 returns the only repo to the pool
4) Timer thread runs (resuming`TryDropARepo`) and it removes the only repo from the pool
5) Thread 1 is stuck here waiting on `TryTake`

For what it's worth I think we could theoretically hit a similar problem if `size` > 1, but that would only be if the thread pool were so backed up that multiple timer callbacks were getting queued up (which seems unlikely).

Thoughts?

Reply:
It's probably worth mentioning that a similar race would be possible if `size` were greater than 1 and all of the repos are in use.  

But I think the difference would be that in step 5 (above) Thread 1 would block until another thread returns a repo to the pool rather than blocking indefinitely (unless perhaps `TryDropARepo` crossed perfectly with `ReturnToPool` each time).

Reply:
You are right that this seems problematic. We currently try to avoid the `TryDropARepo` close to other calls by updating the `dueTime` on the timer. (Keep in mind, running `ResetRepoDisposalTimer()` means we will not fire the timer call until it has not been called for `dueTime`.)

It seems we need a number of "waiting" threads, and that number needs to be zero before we drop a repo.

Reply:
> Keep in mind, running ResetRepoDisposalTimer() means we will not fire the timer call until it has not been called for dueTime.

That's a good point, and @jamill pointed out that it's being called form the start of `GetRepoFromPool` as well.  

There still could be a (admittedly unlikely) race, but it would only be if the timer fired *just* before `ResetRepoDisposalTimer` adjusted it.

> It seems we need a number of "waiting" threads, and that number needs to be zero before we drop a repo.

That sounds good, a "waiting" count should prevent this scenario from happening.